### DISCUSSION TITLE
Extend [DoesNotReturn] attribute impact to execution flow analysis

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#7869] [Language Ideas]
The compiler can compile a switch expression that throws an exception directly.
However, the [compiler can not compile a switch expression that calls a method with the [DoesNotReturn] attribute](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwDoARKCAcxgHsBnBKAYzeoGEWAEwCmAQRgQANgE82UNgG5ixAGYtk5ABQA3CBqjkAvOVILyBgDzk8pU+YDU9gJTEA3sXKfKeAJyaDbADuUAhcABbkHl7R7kTR8V5WNkYAfOQARIA8G4Ag++lwUQnRVjgUhmlZ2Tl5BYWeAPqp5AhhyCyB5DDC7QCiAB5cwgAOnCwwmumikoEQsulOcOQA9IvkAMqoXANsbCqokuRcLGCDUJIQIzA10QC+NU5KcV41VH4BwaERV16xtUXWpeUqvlHr9yMUARkckCvgkGmVyAAVFptTTzJYrATHU7nKCjcjCZCtZAw8i3EHke7EMnEADatBYwjYADkWAgAErCBCoZAwAC6xCoADZKAAWRHIwKoxrNVrtTo9fpDC7jSbTWaUohAA=).
 
```cs
using System;
using System.Diagnostics.CodeAnalysis;

for (var i = 0; i < 100; i++)
{
    Console.WriteLine(i switch 
        {
            < 10 => "🌄",
            < 20 => "🌄🌄",
            _ => throw new Exception("Always"), // Successful compilation
        }
    );
    
    Console.WriteLine(i switch 
        {
            < 10 => "🌄",
            < 20 => "🌄🌄",
            _ => Throw(), // Compilation error
        }
    );
}

[DoesNotReturn]
static void Throw() => throw new Exception("Always");
```

I suggest extending [DoesNotReturn] attribute impact to execution flow analysis. This will allow us to compile the code that calls the method that always throws an exception.

```cs
// Call a method with the [DoesNotReturn] attribute
Throw();

// Will be compiled into the following code
Throw();
UnreachableFlowException.Throw(); // Protect from incorrect [DoesNotReturn] attribute. This method is included in the standard library and is well-known to the compiler.
```

This will protect the execution flow from setting the [DoesNotReturn] attribute incorrectly.

### Related discussions
* #7795

**Comment by 333fred** [comment]
As mentioned in the previous discussion, this is covered by https://github.com/dotnet/csharplang/issues/538. `DoesNotReturn` does not impact execution analysis by design, like all other nullability features, and if we wanted to do so, we'd need to actually design a language feature for that. `never` is that feature.

**Reply by DavidArno** [reply]
"*`never` is that feature.*"

Being spoken in the style of Yoda, concerns I have. "Never" feel prophetic at times. 😉

**Comment by AlexRadch** [comment]
> As mentioned in the previous discussion, this is covered by #538. `DoesNotReturn` does not impact execution analysis by design, like all other nullability features, and if we wanted to do so, we'd need to actually design a language feature for that. `never` is that feature.

It is enough to throw an exception after calling a method with the [DoesNotReturn] attribute. This is what I suggest above.
```cs
// Call a method with the [DoesNotReturn] attribute
Throw();

// Will be compiled into the following code
Throw();
UnreachableFlowException.Throw(); // Protect from incorrect [DoesNotReturn] attribute. This method is included in the standard library and is well-known to the compiler.
```

**Reply by siegfriedpammer** [reply]
Still, the meaning of this exception needs to be specified in the language specification and that requires a new language feature, which @333fred calls `never` which would probably be a return-type like `void` from the perspective of the language and could be implemented like you suggested.

**Reply by RikkiGibson** [reply]
I do not think implicitly inserting a 'throw' is viable. One reason that we have 'Throw' methods which only throw and nothing else is that the presence of a throw statement affects inlineability of the method. Basically, it's desirable to be able to inline everything the method is doing, except the actual throwing part, particularly when it is thought to occur only rarely.

**Reply by GabeSchaffer** [reply]
You don't have to put in a literal `throw`; the compiler can ship with (or emit) a throw helper method and instead implicitly insert calls to that method. As I said in https://github.com/dotnet/csharplang/discussions/7795#discussioncomment-7976818, the runtime could even recognize that pattern and optimize it away.

That said, I'm in favor of a `never` type.

**Reply by CyrusNajmabadi** [reply]
my preference is that if anything is going to impact (non-nrt) flow analysis, that it be part of the language.  so `never` would be the way forward for me :)

**Reply by AlexRadch** [reply]
@CyrusNajmabadi  I agree that analyzing the execution flow during the compilation of such methods is the best solution, but it does not protect against possible breaking changes in libraries.

For example, you used such a method from the library, it was marked with an attribute and checked by the compiler, later the library removed this attribute, and the new version does not always throw an exception. In this case, the flow of execution in the program will change.

To protect the application code from breaking changes in libraries, the compiler should throw an additional exception after such methods.

**Reply by AlexRadch** [reply]
> You don't have to put in a literal `throw`; the compiler can ship with (or emit) a throw helper method and instead implicitly insert calls to that method. As I said in [#7795 (reply in thread)](https://github.com/dotnet/csharplang/discussions/7795#discussioncomment-7976818), the runtime could even recognize that pattern and optimize it away.
> 
> That said, I'm in favor of a `never` type.

@GabeSchaffer Thanks for your clarification! I am against the new type when the same thing can be implemented just as well without adding new entities.

```
// Call a method with the [DoesNotReturn] attribute
Throw();

// Will be compiled into the following code
Throw();
UnreachableFlowException.Throw(); // This method is included in the standard library and is well-known to the compiler.
```

I will make this change to my original proposal.

**Reply by siegfriedpammer** [reply]
> when the same thing can be implemented just as well without adding new entities.

I think, if it's a compiler feature, it would be no problem/nice to have dedicated syntax, defining a new type seems to be the easiest part.

If you are OK with custom attribute and a call to a throw helper injected after each call, you might be able to achieve this feature today by using some IL rewriting library like [Fody](https://github.com/Fody) and a Roslyn analyzer to produce diagnostics/errors if something is wrong.

**Reply by GabeSchaffer** [reply]
I don't think IL rewriting will work because the problem is a spurious compiler error.

**Reply by AlexRadch** [reply]
> If you are OK with custom attribute and a call to a throw helper injected after each call, you might be able to achieve this feature today by using some IL rewriting library like [Fody](https://github.com/Fody) and a Roslyn analyzer to produce diagnostics/errors if something is wrong.

Here we write ideas on how to improve C# and its compiler for everyone.

**Reply by siegfriedpammer** [reply]
> Here we write ideas on how to improve C# and its compiler for everyone.

And the members of the language design team have already answered how they would go about changing the language to solve your problem. This discussion should have ended there.

Thank you for reminding me that I should stop reading and interacting with large parts of the discussions section of this repository.

**Reply by 333fred** [reply]
Folks, let's please keep this civil.
